api/src/main/java/io/grpc/Detachable.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * A <i>Detachable</i> encapsulates an object that can be forked with underlying resources
+ * detached and transferred to a new instance. The forked instance takes over the ownership
+ * of resources and is responsible for releasing after use. The forked instance preserves
+ * states of detached resources. Resources can be consumed through the forked instance as if
+ * being continually consumed through the original instance. The original instance discards
+ * states of detached resources and is no longer consumable as if the resources are exhausted.
+ */
+@ExperimentalApi("TODO")
+public interface Detachable {
+
+  /**
+   * Fork a new instance with underlying resources detached from this instance and
+   * transferred to the new instance.
+   *
+   * @throws IllegalStateException if the underlying resources have already been detached.
```

Comment:
How is the caller supposed to avoid this exception? What if detachable isn't supported?

I suspect we'll want to allow detach to be called multiple times, but the later times obvious don't return anything "real". We could define it as throwing when `available() == 0`, but that seems an abuse of available() without KnownLength. The answers here may be more obvious if this interface becomes tied to InputStream.

Reply:
Updated.